Cortex-A/R AArch32 Thumb-2 support

manuargue opened the pull request:
Allow to optionally build and execute Zephyr in Thumb-2 (T32) ISA for Cortex-A/R AArch32, as it tends to use less memory and yield higher performance than with Arm (A32) instructions.

- Introduce the necessary modifications in the hand-written assembler code, keeping compatibility with existing A32 code.
- For testing purposes, add a new board `fvp_baser_aemv8r_aarch32_thumb` that selects `CONFIG_ISA_THUMB2`
  > [!NOTE]
  > Ideally this could be done through a snippet but `CONFIG_ISA_THUMB2` symbol is hidden and cannot be set through a Kconfig overlay. Any suggestions are welcome.
- Boot-up and build in Thumb-2 mode for NXP S32Z/E devices (Cortex-R52).

manuargue-nxp committed:
nxp_s32: s32ze: execute in Thumb-2 mode by default

Boot-up and build the NXP S32Z/E devices in Thumb-2 mode.

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>

manuargue-nxp committed:
fvp_baser_aemv8r_aarch32: add Thumb-2 board variant

For testing purposes, a new board `fvp_baser_aemv8r_aarch32_thumb`
is introduced that builds and executes in Thumb-2. Ideally this
could be done through a snippet but `CONFIG_ISA_THUMB2` symbol
is hidden and cannot be set through a Kconfig overlay.

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>

povergoing commented on arch/arm/core/aarch32/cortex_a_r/reset.S:
```diff
@@ -201,27 +204,57 @@ EL1_Reset_Handler:
      */
 
     /* FIQ mode stack */
+#if CONFIG_ASSEMBLER_ISA_THUMB2
```
I wonder if we can remove the `#if CONFIG_ASSEMBLER_ISA_THUMB2` and just rewrite with
```
mov r0, #(MODE_FIQ | I_BIT | F_BIT)
msr CPSR_c, r0
```
regardless of THUMB2. There are really too many `#if` defines.

Also, I wonder what will happen if we are going to execute `it ne` in non-Thumb2 mode? Ignored or trap to instruction under exception? The reason I ask this question is just that I think there should be less #if define macros

manuargue replied:
> I wonder if we can remove the #if CONFIG_ASSEMBLER_ISA_THUMB2 and just rewrite with ... regardless of THUMB2. There are really too many #if defines.

yes, it can be done. It will be perhaps redundant for A32 and somebody later may try to "optimize" it. But as long as we have some board in CI building in T32 we can prevent any situation like that.

>  I wonder what will happen if we are going to execute it ne in non-Thumb2 mode?

My understanding is that `IT` is a pseudo-instruction that does not generate any code in A32. From armasm https://developer.arm.com/documentation/dui0473/m/arm-and-thumb-instructions/it . With Zephyr SDK GCC, there's indeed no code generated for the `IT` when in A32. So we could also remove the guarding here.

povergoing replied:
Sounds perfect, and if you worry about someone "optimize" it, then we could have comments telling everyone we need this because of THUMB2. What's your opinion?

povergoing replied:
And we could wait for a while, in case there is a strong objection:)

manuargue replied:
hi @microbuilder @stephanosio could please take a look to this patch and consider above suggestions? thanks

microbuilder replied:
> hi @microbuilder @stephanosio could please take a look to this patch and consider above suggestions? thanks 

Sorry for the slow reply. I'm on vacation this week, and this needs a bit of time to look at properly. 😬 I'll have a look at it as soon as I'm back in the office, though.

microbuilder replied:
> Sounds perfect, and if you worry about someone "optimize" it, then we could have comments telling everyone we need this because of THUMB2. What's your opinion?

I think just adding an appropriate comment is fine here for now, but it's better to leave a signpost about what's going on for future users.

manuargue replied:
guards removed whenever possible

SgrrZhf commented:
It seems that `tests/ztest/error_hook/testing.ztest.error_hook` can not pass on the `fvp_baser_aemv8r_aarch32_thumb` board.

manuargue-nxp committed:
tests: error_hook: skip divide-by-zero fault for Armv8-R

For ARMv8-R, divide by zero trapping is not supported in hardware.

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>

manuargue commented:
> It seems that `tests/ztest/error_hook/testing.ztest.error_hook` can not pass on the `fvp_baser_aemv8r_aarch32_thumb` board.

Divide by zero trapping is not supported for Armv8-R, fixed in https://github.com/zephyrproject-rtos/zephyr/pull/61594/commits/5b63a031cf87d8805be403fb9eef5b90595f62b8

manuargue commented:
@SgrrZhf considering https://github.com/zephyrproject-rtos/zephyr/pull/60031 has two approvals, I think I'll wait for your pr to get merged first and then rebase this one

manuargue-nxp committed:
arm: cortex_r: support executing in Thumb-2

Allow to optionally build and execute Zephyr in Thumb-2 (T32) ISA
for Cortex-A/R AArch32, as it tends to use less memory and yield
higher performance than with Arm (A32) instructions.

Introduce the necessary modifications in the hand-written
assembler code, keeping compatibility with existing A32 code.

Signed-off-by: Manuel Argüelles <manuel.arguelles@nxp.com>

microbuilder commented:
> @SgrrZhf considering #60031 has two approvals, I think I'll wait for your pr to get merged first and then rebase this one

It should be safe to rebase now.

povergoing commented on arch/arm/core/cortex_a_r/exc.S:
```diff
@@ -47,7 +47,11 @@ GTEXT(z_arm_data_abort)
 	 */
 	srsdb sp!, #\mode
+#if defined(CONFIG_ASSEMBLER_ISA_THUMB2)
+	stmfd sp, {r0-r3, r12, lr}
```
TBH, I am actually a bit confused about this since the ^ is not related to THUMB2?

povergoing replied:
Oh, is that because THUMB2 doesn't support the ^ option? I am thinking there is any way that we can also get rid of the guard too?

manuargue replied:
that's right, is not supported. I will check if it's possible

manuargue replied:
hi @povergoing. The `^` seems to be redundant in all the cases used here. When in ARM mode, according to https://developer.arm.com/documentation/dui0489/c/arm-and-thumb-instructions/memory-access-instructions/ldm-and-stm and inspecting the registers e.g. during undef exception handler, there seems to be no effect. I'd say is safe to drop the `^` for both ARM and Thumb2 modes.
But since I'm not so sure why is there in the first place, I want to confirm with you / @microbuilder before doing the change.

manuargue commented:
hi @povergoing , @microbuilder , @SgrrZhf. could you please support me to revisit these changes? The only outstanding comment is https://github.com/zephyrproject-rtos/zephyr/pull/61594#discussion_r1326947394. Thanks

microbuilder commented:
@manuargue Have you tested this against the `qemu_cortex_r5` target, so that we have a CI-friendly target rather than the FVP images, which we don't currently run in CI?

manuargue commented:
> @manuargue Have you tested this against the `qemu_cortex_r5` target, so that we have a CI-friendly target rather than the FVP images, which we don't currently run in CI?

I haven't found a way to set `TEINIT=1` as done for FVP in https://github.com/zephyrproject-rtos/zephyr/blob/568a1501a19d430ba6084add377fbceca0b8749c/boards/arm/fvp_baser_aemv8r_aarch32/board.cmake#L45. Otherwise it won't be possible to enable Thumb-2.
Besides FVP, I've run `tests/kernel` `tests/arch` on real hw (R52) and is all passing.

microbuilder commented:
> > @manuargue Have you tested this against the `qemu_cortex_r5` target, so that we have a CI-friendly target rather than the FVP images, which we don't currently run in CI?
> 
> I haven't found a way to set `TEINIT=1` as done for FVP in
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/568a1501a19d430ba6084add377fbceca0b8749c/boards/arm/fvp_baser_aemv8r_aarch32/board.cmake#L45
> . Otherwise it won't be possible to enable Thumb-2.
> Besides FVP, I've run `tests/kernel` `tests/arch` on real hw (R52) and is all passing.

@pm215 Is there a config flag for QEMU to enable THUMB-2 support on R/A cores coming out of reset, similar to `TEINIT` with the FVP?

manuargue commented:
hi @microbuilder I've run some tests locally with qemu_cortex_r5 and setting the SCTLR.TE bit from the code, but I run into some issues in user-space that I need to check, when running with picolibc built-in Zephyr SDK which was probably compiled for Arm mode. I wasn't able to reproduce the same with R52 hw (Armv8-R), only on the R5 (Armv7-R).

Since I don't think I will be able to sort this out before feature freeze tomorrow due to other topics that demand my attention, I'll leave this to finish it after the freeze.

microbuilder commented:
> Since I don't think I will be able to sort this out before feature freeze tomorrow due to other topics that demand my attention, I'll leave this to finish it after the freeze.

I'd really like to have something in standard CI to test this, sorry if that means missing the freeze. 😬

manuargue commented:
> > Since I don't think I will be able to sort this out before feature freeze tomorrow due to other topics that demand my attention, I'll leave this to finish it after the freeze.
> 
> I'd really like to have something in standard CI to test this, sorry if that means missing the freeze. 😬

Np, I understand and I also think it should be in CI. At the beginning I thought the VFP models are run in CI but later found out is not supported by twister.

pm215 commented:
> > > @manuargue Have you tested this against the `qemu_cortex_r5` target, so that we have a CI-friendly target rather than the FVP images, which we don't currently run in CI?
> > 
> > 
> > I haven't found a way to set `TEINIT=1` as done for FVP in
> > https://github.com/zephyrproject-rtos/zephyr/blob/568a1501a19d430ba6084add377fbceca0b8749c/boards/arm/fvp_baser_aemv8r_aarch32/board.cmake#L45
> > 
> > . Otherwise it won't be possible to enable Thumb-2.
> > Besides FVP, I've run `tests/kernel` `tests/arch` on real hw (R52) and is all passing.
> 
> @pm215 Is there a config flag for QEMU to enable THUMB-2 support on R/A cores coming out of reset, similar to `TEINIT` with the FVP?

We don't currently have an equivalent of the TEINIT config signal on the real Cortex-R5, because we don't have a board model where the real hardware/SoC would wire TEINIT high. It would be easy to implement, but there's no current use for it.

In any case, all it does is set the reset value of SCTLR.TE, so you can have a trivial piece of Arm code at your reset vector that sets it and then switches to Thumb mode if you want to have your OS be mostly Thumb2. That's probably more useful anyway, because my guess is most real hardware wires TEINIT low.

manuargue commented:
closing for now as there have been too many changes in arm/ and due to other priorities I couldn't come back to this after last release. To come back with a rebase and remaining QEMU Cortex-R5 support later on.
